orttraining/orttraining/core/framework/distributed_run_context.cc

```diff
@@ -47,14 +45,46 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
   params_.data_parallel_size = data_parallel_size;
   params_.horizontal_parallel_size = horizontal_parallel_size;
   params_.pipeline_stage_size = pipeline_stage_size;
-  groups_.resize(2);
+  groups_.resize(static_cast<size_t>(WorkerGroupType::WorkerGroupTypeCount));
 
+  // Consider distributed training forms three axis, data parallel, horizontal parallel and pipeline parallel.
+  // The three axis are pependicular to each other and like x, y, z axis in 3D space (but only the positive directions).
+  // Now the world ranks are numbers filling in this 3D space alone the three axis. It will try fill alone the
+  // horizontal axis first, then data parallel axis, and last pipeline parallel axis.
+  //
+  // Ranks that are aligned with a specific axis forms a group. For example, for a 3D cube with size
+  // horizontal_parallel x data_parallel x pipeline_parallel equal to 4x3x2, there are 24 rank numbers fill into the
+  // cubic. It will have 6 horizontal groups, 8 data parallel groups and 12 pipeline groups, as shown below.
+  //
+  //          pipeline
+  //            ^
```

Comment:
For machine ID c, are the following equations correct?

x (in_group_hori_group_id) = c % 4
y (in_group_data_group_id) = (c / 4) % 3
z (in_group_pipe_group_id) = (c / 12) % 2

To recover linear index x, we can use x + y * 4 + z * 4 * 3.
To see which hori group x belongs to, we can reuse the linear-index equation but assume there is no hori axis by
 1. set dimension of hori group to 1 (instead of 4)
 2. set  in_group_hori_group_id to 0.
hori_group_id = 0 +y * 1 +z * 1 * 3
Similarly, we have
data_group_id = x + 0 * 3 + z * 4 * 1
pipe_group_id = x + y * 4 + 0 * 4 * 3

A more systematic understanding is to think this as calculating linear index in the projected space.

Reply:
This is neat. Thanks. Update. 

---
In reply to: [478766878](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r478766878) [](ancestors = 478766878)